src/framework/mlt_chain.c

```diff
@@ -0,0 +1,356 @@
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include "mlt_chain.h"
+#include "mlt_frame.h"
+#include "mlt_log.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+
+/** \brief private service definition */
+
+typedef struct
+{
+	int link_count;
+	int link_size;
+	mlt_link* links;
+	mlt_producer source;
+	mlt_profile source_profile;
+	mlt_producer begin;
+}
+mlt_chain_base;
+
+/* Forward references to static methods.
+*/
+
+static int producer_get_frame( mlt_producer parent, mlt_frame_ptr frame, int index );
+static void relink_chain( mlt_chain self );
+
+/** Construct a link.
+ *
+ * Sets the resource property to "<link>", the mlt_type to "mlt_producer",
+ * and mlt_service to "link".
+ *
+ * \public \memberof mlt_chain_s
+ * \return the new link
+ */
+
+mlt_chain mlt_chain_init( mlt_profile profile )
+{
+	mlt_chain self = calloc( 1, sizeof( struct mlt_chain_s ) );
+	if ( self != NULL )
+	{
+		mlt_producer producer = &self->parent;
+		if ( mlt_producer_init( producer, self ) == 0 )
+		{
+			mlt_properties properties = MLT_PRODUCER_PROPERTIES( producer );
+
+			mlt_properties_set( properties, "resource", "<chain>" );
+			mlt_properties_set( properties, "mlt_type", "mlt_producer" );
+			mlt_properties_set( properties, "mlt_service", "chain" );
+
+			producer->get_frame = producer_get_frame;
+			producer->close = ( mlt_destructor )mlt_chain_close;
+			producer->close_object = self;
+
+			mlt_service_set_profile( MLT_CHAIN_SERVICE( self ), profile );
+
+			// Generate local space
+			self->local = calloc( 1, sizeof( mlt_chain_base ) );
+			mlt_chain_base* base = self->local;
+			base->source_profile = mlt_profile_init(NULL);
+		}
+		else
+		{
+			free( self );
+			self = NULL;
+		}
+	}
+	return self;
+}
+
+/** Set the source producer.
+ *
+ * \public \memberof mlt_chain_s
+ * \param self a chain
+ * \param source the new source producer
+ */
+
+void mlt_chain_set_source( mlt_chain self, mlt_producer source )
+{
+	int error = self == NULL || source == NULL;
+	if ( error == 0 )
+	{
+		mlt_chain_base* base = self->local;
+
+		// Save the source producer
+		mlt_producer_close( base->source );
+		base->source = source;
+		mlt_properties_inc_ref( MLT_PRODUCER_PROPERTIES( source ) );
+
+		// Set the producer to be in native frame rate
+		mlt_profile_from_producer( base->source_profile, base->source );
+		mlt_service_set_profile( MLT_PRODUCER_SERVICE(base->source), base->source_profile );
```

Comment:
The real source producer for the entire chain is configured to be in a "native" profile so that the next link has access to every single frame. Some producer implementations may need to be updated to work better when the profile is changed. For example, producer_avformat works fine except that the length property is not updated after the profile framerate changes.